File: souq-local/backend/app/services/service_pricing.py
"""Service pricing model validation and normalization."""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, Field, ValidationError, model_validator


class PricingModel(str, Enum):
    FIXED_PRICE = "fixed_price"
    STARTING_FROM = "starting_from"
    PRICE_RANGE = "price_range"
    HOURLY = "hourly"
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    PER_PERSON = "per_person"
    PER_UNIT = "per_unit"
    PER_SQM = "per_sqm"
    PER_KM = "per_km"
    REQUEST_QUOTE = "request_quote"
    CONTACT_FOR_PRICE = "contact_for_price"
    NEGOTIABLE = "negotiable"
    FREE = "free"


_SINGLE_PRICE_MODELS = {
    PricingModel.FIXED_PRICE,
    PricingModel.STARTING_FROM,
    PricingModel.HOURLY,
    PricingModel.DAILY,
    PricingModel.WEEKLY,
    PricingModel.MONTHLY,
    PricingModel.PER_PERSON,
    PricingModel.PER_UNIT,
    PricingModel.PER_SQM,
    PricingModel.PER_KM,
}

_NO_PRICE_MODELS = {
    PricingModel.REQUEST_QUOTE,
    PricingModel.CONTACT_FOR_PRICE,
}


class ServicePricingFields(BaseModel):
    pricing_model: PricingModel = PricingModel.FIXED_PRICE
    price_mad: float | None = Field(default=None, ge=0, le=10_000_000)
    price_min_mad: float | None = Field(default=None, ge=0, le=10_000_000)
    price_max_mad: float | None = Field(default=None, ge=0, le=10_000_000)
    price_negotiable: bool = False

    @model_validator(mode="after")
    def validate_pricing(self) -> ServicePricingFields:
        model = self.pricing_model

        if model in _NO_PRICE_MODELS:
            self.price_mad = None
            self.price_min_mad = None
            self.price_max_mad = None
            self.price_negotiable = False
            return self

        if model == PricingModel.FREE:
            self.price_mad = 0.0
            self.price_min_mad = None
            self.price_max_mad = None
            self.price_negotiable = False
            return self

        if model == PricingModel.NEGOTIABLE:
            self.price_negotiable = True
            self.price_min_mad = None
            self.price_max_mad = None
            return self

        if model == PricingModel.PRICE_RANGE:
            if self.price_min_mad is None or self.price_max_mad is None:
                raise ValueError("Price range requires both minimum and maximum prices")
            if self.price_min_mad > self.price_max_mad:
                raise ValueError("Minimum price cannot exceed maximum price")
            self.price_mad = None
            self.price_negotiable = False
            return self

        if model in _SINGLE_PRICE_MODELS:
            if self.price_mad is None:
                raise ValueError("This pricing model requires a price")
            self.price_min_mad = None
            self.price_max_mad = None
            self.price_negotiable = False
            return self

        return self


def infer_pricing_model(
    *,
    price_mad: float | None,
    price_negotiable: bool,
    pricing_model: str | None = None,
) -> PricingModel:
    if pricing_model:
        return PricingModel(pricing_model)
    if price_negotiable:
        return PricingModel.NEGOTIABLE
    if price_mad is None:
        return PricingModel.REQUEST_QUOTE
    if price_mad == 0:
        return PricingModel.FREE
    return PricingModel.FIXED_PRICE


def normalize_service_pricing(data: dict) -> dict:
    validated = ServicePricingFields.model_validate(data)
    return validated.model_dump()

File: souq-local/mobile/lib/core/models/service_pricing.dart
import '../../l10n/strings/app_strings.dart';

enum ServicePricingModel {
  fixedPrice('fixed_price'),
  startingFrom('starting_from'),
  priceRange('price_range'),
  hourly('hourly'),
  daily('daily'),
  weekly('weekly'),
  monthly('monthly'),
  perPerson('per_person'),
  perUnit('per_unit'),
  perSqm('per_sqm'),
  perKm('per_km'),
  requestQuote('request_quote'),
  contactForPrice('contact_for_price'),
  negotiable('negotiable'),
  free('free');

  const ServicePricingModel(this.apiValue);

  final String apiValue;

  static ServicePricingModel fromApi(String? value) {
    return ServicePricingModel.values.firstWhere(
      (model) => model.apiValue == value,
      orElse: () => ServicePricingModel.fixedPrice,
    );
  }

  bool get requiresSinglePrice =>
      this == fixedPrice ||
      this == startingFrom ||
      this == hourly ||
      this == daily ||
      this == weekly ||
      this == monthly ||
      this == perPerson ||
      this == perUnit ||
      this == perSqm ||
      this == perKm ||
      this == negotiable;

  bool get requiresPriceRange => this == priceRange;

  bool get hidesPriceInput =>
      this == requestQuote || this == contactForPrice || this == free;

  String label(AppStrings l10n) {
    switch (this) {
      case ServicePricingModel.fixedPrice:
        return l10n.pricingModelFixedPrice;
      case ServicePricingModel.startingFrom:
        return l10n.pricingModelStartingFrom;
      case ServicePricingModel.priceRange:
        return l10n.pricingModelPriceRange;
      case ServicePricingModel.hourly:
        return l10n.pricingModelHourly;
      case ServicePricingModel.daily:
        return l10n.pricingModelDaily;
      case ServicePricingModel.weekly:
        return l10n.pricingModelWeekly;
      case ServicePricingModel.monthly:
        return l10n.pricingModelMonthly;
      case ServicePricingModel.perPerson:
        return l10n.pricingModelPerPerson;
      case ServicePricingModel.perUnit:
        return l10n.pricingModelPerUnit;
      case ServicePricingModel.perSqm:
        return l10n.pricingModelPerSqm;
      case ServicePricingModel.perKm:
        return l10n.pricingModelPerKm;
      case ServicePricingModel.requestQuote:
        return l10n.pricingModelRequestQuote;
      case ServicePricingModel.contactForPrice:
        return l10n.pricingModelContactForPrice;
      case ServicePricingModel.negotiable:
        return l10n.pricingModelNegotiable;
      case ServicePricingModel.free:
        return l10n.pricingModelFree;
    }
  }
}

String formatServicePrice(
  AppStrings l10n, {
  required ServicePricingModel pricingModel,
  double? priceMad,
  double? priceMinMad,
  double? priceMaxMad,
  bool priceNegotiable = false,
}) {
  String amount(double value) {
    final decimals = value % 1 == 0 ? 0 : 2;
    return '${value.toStringAsFixed(decimals)} MAD';
  }

  switch (pricingModel) {
    case ServicePricingModel.fixedPrice:
      return priceMad == null ? l10n.priceOnRequest : amount(priceMad);
    case ServicePricingModel.startingFrom:
      return priceMad == null ? l10n.priceOnRequest : l10n.priceStartingFrom(amount(priceMad));
    case ServicePricingModel.priceRange:
      if (priceMinMad == null || priceMaxMad == null) return l10n.priceOnRequest;
      return l10n.priceRangeLabel(amount(priceMinMad), amount(priceMaxMad));
    case ServicePricingModel.hourly:
      return priceMad == null ? l10n.priceOnRequest : l10n.pricePerHour(amount(priceMad));
    case ServicePricingModel.daily:
      return priceMad == null ? l10n.priceOnRequest : l10n.pricePerDay(amount(priceMad));
    case ServicePricingModel.weekly:
      return priceMad == null ? l10n.priceOnRequest : l10n.pricePerWeek(amount(priceMad));
    case ServicePricingModel.monthly:
      return priceMad == null ? l10n.priceOnRequest : l10n.pricePerMonth(amount(priceMad));
    case ServicePricingModel.perPerson:
      return priceMad == null ? l10n.priceOnRequest : l10n.pricePerPerson(amount(priceMad));
    case ServicePricingModel.perUnit:
      return priceMad == null ? l10n.priceOnRequest : l10n.pricePerUnit(amount(priceMad));
    case ServicePricingModel.perSqm:
      return priceMad == null ? l10n.priceOnRequest : l10n.pricePerSqm(amount(priceMad));
    case ServicePricingModel.perKm:
      return priceMad == null ? l10n.priceOnRequest : l10n.pricePerKm(amount(priceMad));
    case ServicePricingModel.requestQuote:
      return l10n.pricingModelRequestQuote;
    case ServicePricingModel.contactForPrice:
      return l10n.pricingModelContactForPrice;
    case ServicePricingModel.negotiable:
      if (priceMad != null) return l10n.priceNegotiableWithAmount(amount(priceMad));
      return l10n.priceNegotiable;
    case ServicePricingModel.free:
      return l10n.pricingModelFree;
  }
}
